man/jpeghotp.1.in
.\" This file is part of jpegpixi, a program to interpolate pixels in
.\" JFIF image files.
.\" Copyright (C) 2003, 2004, 2005 Martin Dickopp
.
.\" Jpegpixi is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published
.\" by the Free Software Foundation; either version 2 of the License,
.\" or (at your option) any later version.
.
.\" Jpegpixi is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.
.\" You should have received a copy of the GNU General Public License
.\" along with jpegpixi; if not, write to the Free Software
.\" Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301,
.\" USA.
.
.TH JPEGHOTP 1 18-Apr-2005 "@PACKAGE_STRING@" "User Commands"
.SH NAME
jpeghotp \- find hot or dead pixels in otherwise black or white JPEG images
.SH SYNOPSIS
.B jpeghotp
.RI [ OPTION ]...\&
.I JPEG-FILE
.RI [ PIXEL-BLOCKS-FILE ]
.SH DESCRIPTION
Jpeghotp automatically finds hot pixels in an otherwise black JPEG image
(default) or dead pixels in an otherwise white JPEG image (if the
.B \-\-invert
option is specified). The output is suitable as a pixel block
specification file for the
.BR jpegpixi (1)
program. A typical application is to process an image taken with a digital
camera with CCD defects.
.PP
To obtain a black image, take a picture in total darkness or with
covered lens. To obtain a while image, take an overexposed picture of a
well-lit piece of white paper.
.PP
.I JPEG-FILE
is the filename of the JPEG image to read. It can be
.BR \- ,
which refers to standard input.
.PP
.I PIXEL-BLOCKS-FILE
is the name of the file to write the coordinates and sizes of the hot
(or dead) pixel blocks to. If it is omitted, the coordinates and sizes
are written to standard output.
.SH OPTIONS
.IP "\fB\-t\fR \fITHRESHOLD\fR, \fB\-\-threshold\fR=\fITHRESHOLD\fR"
Specify the threshold between noise and signal. Pixel luminosities below
the threshold are considered CCD noise, luminosities above are considered
hot pixels.
.I THRESHOLD
must be a number between 0 and 100 followed by a percent character
.RB ( % ).
The default is
.BR 10% .
.IP "\fB\-i\fr, \fB\-\-invert\fR"
Invert the image. If this option is specified, dead pixels in an
otherwise white image instead of hot pixels in an otherwise black image
are found.
.IP "\fB\-c\fR, \fB\-\-comments\fR"
Output comments with the luminosities of the hot (or dead) pixel blocks.
This can be helpful in hand tuning the pixel blocks file.
.IP "\fB\-\-help\fR"
Display a short help text and exit immediately.
.IP "\fB\-\-version\fR"
Display version information and exit immediately.
.SH BUGS
Please report bugs to <martin\-jpegpixi@zero\-based.org>.
.SH COPYRIGHT
Copyright \(co 2003, 2004, 2005 Martin Dickopp
.PP
Jpeghotp is free software; it may be copied and/or modified under the
terms of the GNU General Public License version 2 or (at your option)
any later version. There is NO warranty; not even for MERCHANTABILITY
or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
.BR jpegpixi (1)
.PP
The jpegpixi homepage:
http://www.zero\-based.org/software/jpegpixi/
.
.\" Local Variables:
.\" mode: nroff
.\" coding: iso-8859-1
.\" End:


NEWS
Version 1.1.1:
* Fixed typo in English jpegpixi manual page.
* Updated postal address of the Free Software Foundation.

Version 1.1.0:
* In addition to hot pixels in an otherwise black image, the `jpeghotp'
  program can now also find dead pixels in an otherwise white image.
* Fail gracefully if the locale cannot be set.

Version 1.0.3:
* Stephan Peccini <stephan@tesenca.info> has contributed French
  translations of the manual pages and the message catalog. Thank
  you, Stephan!

Version 1.0.2:
* Spelling mistakes in the German message catalog have been corrected.
* Minor cleanup of the code and build infrastructure. Split some
  messages to make the job of translators easier.

Version 1.0.1:
* Jarno Elonen <elonen@iki.fi> has contributed Finnish translations of
  the manual pages and the message catalog. Thank you, Jarno!

Version 1.0.0:
* Native Language Support is provided. The package contains a message
  catalog with German translations.
* With Native Language Support, all intended features are implemented.
  Therefore, the major version number is changed to 1.

Version 0.16.0:
* German translations of the manual pages are provided. This is the
  first step towards Native Language Support in the programs themselves.

Version 0.15.1:
* Following a proposal on the debian-legal mailinglist (see
  <http://lists.debian.org/debian-legal/2004/05/msg00150.html> and
  followups), license information (i.e. the output shown if the
  `--version' command line option is specified) is put into a separate
  section `.license' in the executable, if possible.  The contents of
  this section can be dumped with the GNU binutils, using the command:
  objcopy -O binary -j .license /path/to/jpegpixi `tty`
* Bug fix: Link with math library (-lm) if necessary.

Version 0.15.0:
* The `jpegpixi' program has a new command line option `--verbose' to
display coordinates and sizes of interpolated pixel blocks.
* Bug fix: Warn if overlapping pixel blocks are specified.
* Bug fix: Consolidate adjacent pixel blocks. For example, if `42,17'
and `43,17' are specified, the behavior is the same as if `42,17,2,1'
had been specified instead.
* Bug fix: Work correctly in extreme cases where almost the whole image
is interpolated.

Version 0.14.2:
* The `jpeghotp' program can output comments with the luminosities of
the hot pixel blocks. This can be helpful in "hand tuning" the pixel
blocks file.

Version 0.14.1:
* Jarno Elonen <elonen@iki.fi> will maintain the jpegpixi Debian package;
therefore the `debian' directory has been removed.

Version 0.14:
* A new program `jpeghotp' is provided. It finds hot pixels in an
otherwise black JPEG image.
* Spelling mistakes in the manual page corrected.

Version 0.13:
* Coordinates and sizes can not only be specified as absolute values, but
also as percentages of the image size.

Version 0.12:
* Jpegpixi now parses command line options with code generated with `opag',
an option parser generator by the same author. This has no user-visible effect.

Version 0.11:
* No code changes at all. The only purpose of this release is to incorporate
a change of the home page and the author's e-mail address.

Version 0.10:
* Pixel block specifications can be read from a file (in addition to
the command line).

Version 0.9:
* In addition to 2-dimensional interpolation, 1-dimensional interpolation is
now also supported. The latter can be used to remove horizontal or vertical
stripes which span the whole image or a large part of it. Thanks to Evgeny
Stambulchik for suggesting this.
* Bug fixes.

Version 0.8:
* The size of the block to be interpolated is now only limited by available
memory.
* Jpegpixi does no longer use precalculated tables for the interpolation;
this reduces the size of the executable.
* Especially for low-quality images, the (inevitable) quality loss in the
vicinity of interpolated pixels is now even smaller. To achieve this, the
optimal threshold for the rounding of DCT coefficients has been empirically
determined.
* Huffman tables are re-calculated for the output image, resulting in a
smaller file size in some cases.
* It is no longer possible to print debugging output. Maintaining this
functionality is too much of a hassle for too little benefit.

Version 0.7:
* Bug fix: Pixels near the image edges are treated correctly.
* More verbose debugging output.
* Uses automake 1.7.5.

Version 0.6:
* The package uses autoconf 2.57 and automake 1.7.2.
* A `debian' subdirectory to build a Debian package is provided.

Version 0.5:
* There is now a manual page.
* The maximum size of the pixel block to be interpolated has been
increased to 5x5 pixels.
* The package uses automake 1.6.1.

Version 0.4:
* Biquadratic and bicubic interpolation are supported in addition to
averaging and bilinear interpolation.

Version 0.3:
* The size of the pixel block to be interpolated can be specified
(up to 3x3 pixels).
* Interpolation engine rewritten from scratch. Averaging and bilinear
interpolation are available as interpolation methods. Quadratic
interpolation has been taken out and will be added again to a later
version. CAUTION: The names of the interpolation methods (`-m' option)
have changed.

Version 0.2:
* Bug fixes. Now jpegpixi works reasonably well with low quality JFIF
images.
* New command line option `-d' to display debugging information (e.g.
the DCT coefficients before and after the interpolation).
* Interpolation method can be specified on the command line. The
following methods are supported:
  - average of 4 adjacent pixels
  - average of 8 surrounding pixels
  - quadratic surface defined by 12 nearest pixels

Version 0.1:
* First release.


jpeghotp.c
/* This file is part of jpegpixi, a program to interpolate pixels in
   JFIF image files.
   Copyright (C) 2003, 2004, 2005 Martin Dickopp

   Jpegpixi is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published
   by the Free Software Foundation; either version 2 of the License,
   or (at your option) any later version.

   Jpegpixi is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with jpegpixi; if not, write to the Free Software
   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301,
   USA.  */

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif
#include "util.h"

#include <stdio.h>

#if HAVE_ERRNO_H
# include <errno.h>
#endif
#ifndef errno
extern int errno;
#endif

#if STDC_HEADERS
# include <ctype.h>
#else
# define isspace(c) ((c) == ' ' || (c) == '\t')
#endif

#if HAVE_NLS
# include <locale.h>
#endif

#include "jpegpixi.h"
#include "rbtree.h"
#include "opthotp.h"


/* Name of this program.  */
#define PROGRAM_NAME "jpeghotp"



/* Display text in response to the --help command line option.  */
extern void display_help_text (void);

/* Display text in response to the --version command line option.  */
extern void display_version_text (void);



/* Input file name (needed by display_error_message).  */
static const char *infilename1;



static void display_error_message (j_common_ptr jpg);
static int point_lumi_s_validate_and_consolidate (void *a, const void *b) gcc_attr_pure gcc_attr_nonnull (());



int
main (int argc, char *argv [])
{
    int i;
    const char *infilename, *outfilename;
    struct jpeg_error_mgr errmgr;
    struct jpeg_decompress_struct injpg;
    FILE *infile, *outfile;
    struct rbtree points;
    unsigned int threshold = DENOM / 10;



    /* Store program invocation name.  */
    invocation_name = argv [0] != 0 && *(argv [0]) != '\0' ? argv [0] : PROGRAM_NAME;


    /* Set up Native Language Support.  */
#if HAVE_NLS
    if (setlocale (LC_ALL, "") != 0)
    {
        if (bindtextdomain (PACKAGE_TARNAME, LOCALEDIR) == 0 || textdomain (PACKAGE_TARNAME) == 0)
            mem_alloc_failed ();
    }
    else
    {
        fprintf (stderr, "%s: cannot set locale\n"
                 "Make sure the `LC_*' and `LANG' environment variables are set to valid values.\n", invocation_name);
        exit (1);
    }
#endif


   /* Parse command line options.  */
    if ((i = parse_options (invocation_name, argc, argv)) == -1)
        goto error_try_help;

    if (opt_help)
    {
        display_help_text ();
        exit (0);
    }

    if (opt_version)
    {
        display_version_text ();
        exit (0);
    }

    if (opt_threshold)
    {
        const char *p = arg_threshold;

        while (isspace ((unsigned char)*p))
            ++p;

        if (parse_number (&p, &threshold) != 0)
        {
          error_threshold:
            fprintf (stderr, _("%s: invalid threshold `%s'\n"), invocation_name, arg_threshold);
            goto error_try_help;
        }

        while (isspace ((unsigned char)*p))
            ++p;
        if (*p != '\0')
            goto error_threshold;
    }


    /* Parse non-option command line arguments.  */
    if (i < argc)
    {
        infilename = argv [i++];
        if (strcmp (infilename, "-") == 0)
            infilename = 0;
    }
    else
    {
        fprintf (stderr, _("%s: source filename missing\n"), invocation_name);
      error_try_help:
        fprintf (stderr, _("Try `%s --help' for more information.\n"), invocation_name);
        exit (1);
    }

    if (i < argc)
    {
        outfilename = argv [i++];

        if (i < argc)
        {
            fprintf (stderr, _("%s: unexpected argument `%s'\n"), invocation_name, argv [i]);
            goto error_try_help;
        }

        if (strcmp (outfilename, "-") == 0)
            outfilename = 0;
    }
    else
        outfilename = 0;


    /* Initialize jpeglib objects.  */
    infilename1 = infilename;

    jpeg_std_error (&errmgr);
    errmgr.output_message = &display_error_message;

    jpeg_create_decompress (&injpg);
    injpg.err = &errmgr;


    /* Open input file, read header, and start decompression.  */
    if (infilename != 0)
    {
        if ((infile = fopen (infilename, "rb")) == 0)
        {
            fprintf (stderr, _("%s: %s: %s\n"), invocation_name, infilename, strerror (errno));
            exit (1);
        }
    }
    else
        infile = stdin;

    jpeg_stdio_src (&injpg, infile);
    jpeg_read_header (&injpg, TRUE);
    jpeg_start_decompress (&injpg);


    /* Find hot pixels and store them in a tree.  */
    {
        unsigned int y;
        JSAMPLE *samples = xmalloc (injpg.output_width * injpg.output_components * sizeof *samples);
        const int threshold_value = REL_TO_ABS (threshold, MAXJSAMPLE);

        rbtree_create (&points);

        for (y = 0; injpg.output_scanline < injpg.output_height; ++y)
        {
            unsigned int x;
            JSAMPLE *s;

            jpeg_read_scanlines (&injpg, &samples, 1);

            for (x = 0, s = samples; x < (unsigned int)injpg.output_width; ++x)
            {
                size_t icomp;
                int value;

                /* Find maximum component value.  Invert sample values if the `--invert' command line option
                   has been specified.  */
                for (icomp = 0, value = 0; icomp < (size_t)injpg.output_components; ++icomp, ++s)
                {
                    const int sample_value = opt_invert ? MAXJSAMPLE - GETJSAMPLE (*s) : GETJSAMPLE (*s);

                    if (sample_value > value)
                        value = sample_value;
                }

                if (value > threshold_value)
                {
                    struct point_lumi_s point;

                    point.p.x = x;
                    point.p.y = y;
                    point.p.x_size = 1;
                    point.p.y_size = 1;
                    point.lumi = ((2 * DENOM) / MAXJSAMPLE) * value / 2;

                    if (rbtree_insert (&points, &point, sizeof point, point_s_cmp, 0) == 0)
                        mem_alloc_failed ();
                }
            }
        }

        free (samples);
        consolidate_pixels (&points, point_lumi_s_validate_and_consolidate);
    }


    /* Finish processing, close input file, and delete jpeglib object.  */
    jpeg_finish_decompress (&injpg);

    if (infilename != 0 && fclose (infile) == EOF)
    {
        fprintf (stderr, _("%s: %s: %s\n"), invocation_name, infilename, strerror (errno));
        exit (1);
    }

    jpeg_destroy_decompress (&injpg);



    /* Write output file.  */
    if (outfilename != 0)
    {
        if ((outfile = fopen (outfilename, "w")) == 0)
        {
            fprintf (stderr, _("%s: %s: %s\n"), invocation_name, outfilename, strerror (errno));
            exit (1);
        }
    }
    else
        outfile = stdout;

    {
        const struct point_lumi_s *point = rbtree_first (&points);

        while (point != 0)
        {
            fprintf (outfile, "%u,%u,%u,%u", point->p.x, point->p.y, point->p.x_size, point->p.y_size);
            if (opt_comments)
            {
                const unsigned int lumi_per_10000 = (point->lumi / DENOM_SQRT) * 10000 / DENOM_SQRT;
                fprintf (outfile, _("\t# luminosity: %u.%02u%%"), lumi_per_10000 / 100, lumi_per_10000 % 100);
            }
            putc ('\n', outfile);
            if (ferror (outfile))
            {
                fprintf (stderr, _("%s: %s: %s\n"), invocation_name, outfilename != 0 ? outfilename : _("STDOUT"),
                         strerror (errno));
                exit (1);
            }

            point = rbtree_next (point);
        }
    }

    if (outfilename != 0 && fclose (outfile) == EOF)
    {
        fprintf (stderr, _("%s: %s: %s\n"), invocation_name, outfilename, strerror (errno));
        exit (1);
    }


    exit (0);
}



/* Display an error message.  */
static void
display_error_message (const j_common_ptr jpg)
{
    const char *filename = infilename1 != 0 ? infilename1 : _("STDIN");
    char buffer [JMSG_LENGTH_MAX];

    (*jpg->err->format_message) (jpg, buffer);
    fprintf (stderr, _("%s: %s: %s\n"), invocation_name, filename, buffer);
}



/* Validate and consolidate two point structures.  The validation always succeeds.  */
static int
point_lumi_s_validate_and_consolidate (void *const a, const void *const b)
{
    struct point_lumi_s *const point_a = a;
    const struct point_lumi_s *const point_b = b;

    point_a->lumi = MAX (point_a->lumi, point_b->lumi);
    return 1;
}
